sdk/pinocchio/src/sysvars/slot_hashes/mod.rs

```diff
@@ -0,0 +1,345 @@
+}
+
+/// Validates SlotHashes data format assuming golden mainnet length and returns the entry count.
+///
+/// The function checks:
+/// 1. The buffer length is exactly `MAX_SIZE` bytes.
+/// 2. The declared entry count is ≤ `MAX_ENTRIES`.
+///
+/// It returns `Ok(())` if the data is well-formed, otherwise an appropriate
+/// `ProgramError` describing the issue.
+#[inline]
+fn parse_and_validate_data(data: &[u8]) -> Result<(), ProgramError> {
+    // Must be exactly the golden mainnet size
+    if data.len() != MAX_SIZE {
+        return Err(ProgramError::Custom(ERR_DATA_LEN_MISMATCH));
+    }
+
+    // Read and validate the entry count from the header
+    let num_entries = read_entry_count_from_bytes(data).ok_or(ProgramError::AccountDataTooSmall)?;
+
+    // num_entries < 512 is allowed, for contexts which padded data up to size
+    if num_entries > MAX_ENTRIES {
+        return Err(ProgramError::Custom(ERR_ENTRYCOUNT_OVERFLOW));
+    }
+    Ok(())
+}
+
+impl SlotHashEntry {
+    /// Returns the slot number as a u64.
+    #[inline(always)]
+    pub fn slot(&self) -> Slot {
+        u64::from_le_bytes(self.slot_le)
+    }
+}
+
+impl<T: Deref<Target = [u8]>> SlotHashes<T> {
+    /// Creates a `SlotHashes` instance from mainnet-sized data with full validation.
+    ///
+    /// This constructor expects exactly MAX_SIZE (20,488) bytes and performs validation
+    /// of the entry count. Callers with different buffer sizes must pad their data
+    /// to the required length or use test-specific constructors.
+    /// Does not validate that entries are sorted in descending order.
+    #[inline(always)]
+    pub fn new(data: T) -> Result<Self, ProgramError> {
+        parse_and_validate_data(&data)?;
+        // SAFETY: `parse_and_validate_data` verifies that the data slice is exactly
+        // `MAX_SIZE` bytes long and that the declared entry count is within
+        // `MAX_ENTRIES`, thus upholding all invariants required by
+        // `SlotHashes::new_unchecked`.
+        Ok(unsafe { Self::new_unchecked(data) })
+    }
+
+    /// Creates a `SlotHashes` instance assuming golden mainnet buffer size.
+    /// Reads the entry count from the data but assumes the buffer is MAX_SIZE bytes.
+    /// Callers with different needs must pad their incomplete sysvar data up to
+    /// the required length in test or new network contexts.
+    ///
+    /// # Safety
+    ///
+    /// This function is unsafe because it does not validate the data size or format.
+    /// The caller must ensure:
+    /// 1. The underlying byte slice in `data` represents valid SlotHashes data
+    ///    (length prefix + entries, where entries are sorted in descending order by slot).
+    /// 2. The data slice contains exactly MAX_SIZE (20,488) bytes.
+    /// 3. The first 8 bytes contain a valid entry count in little-endian format.
+    ///
+    #[inline(always)]
+    pub unsafe fn new_unchecked(data: T) -> Self {
+        debug_assert!(data.len() == MAX_SIZE);
```

Comment:
Not sure if this matches the `_unchecked()` API design.
But you can also call `parse_and_validate_data(&data)?` here in the debug build:

```suggestion
        if cfg!(debug_assertions) {
            parse_and_validate_data(&data)
                .expect("`data` matches all the same requirements as for `new()`");
        }
```

If you are already doing some checks in debug mode, why not run all of them?